### ISSUE TITLE
Add a new filter option $message for strict blocking

### ISSUE THREAD

**Issue by Yuki2718** [uBlockOrigin/uBlock-issues#1195]
### Prerequisites

- [x] I verified that this is not a filter issue
    - Filter issues MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues)
    
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uBlockOrigin](https://old.reddit.com/r/uBlockOrigin/)
    
- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
- I tried to reproduce the issue when...
    - [x] uBlock Origin is the only extension
    - [x] uBlock Origin with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uBlock Origin
- [x] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### Description

I'm impressed with clever use of `$domain=` by @DandelionSprout https://github.com/DandelionSprout/adfilt/pull/81#issuecomment-655701790, however, in this case you can't use `$domain` for its original purpose. This can be significant disadvantage as those rules for bad domains can well be untokenizable regex rules which should be narrowed down by `$domain`. So my proposal here, add `$message` (or whatsoever suitable English word) to show messages on the strict-blocked page.
Ex.
`||badsite.com^$message="The site is blocked for phishing activity!"`

Correct me if I'm wrong but I believe this option can and should be ignored on filter parsing - strict-blocked page shows original list name, which means reverse-lookup was done so why doesn't fetch the value of this new option at this step. I wonder if @curbengh, in addition to @DandelionSprout of course, has interest in using this if it's adopted. Other than bad domains, I will use this option for links with tracking parameter. Most users of my lists would have no idea about what to do if such a link is blocked and all my experiences tell that explaining that (i.e. clicking a loupe icon to get a non-tracking link) in README is far from enough.

### A specific URL where the issue occurs

Really any URL that can be strict-blocked, but e.g. `http://rsalcau.com/` as it's blocked by EL.

### Steps to Reproduce

Skipping

### Expected behavior:

Messages can be shown on strict-blocked page

### Actual behavior:

There's no way other than exploiting `$domain=`.

### Your environment

* uBlock Origin version: 1.28.4
* Browser Name and version: Firefox 79.0
* Operating System and version: Windows 10

**Comment by DandelionSprout** [comment]
I see what you're thinking of, and I thank you for the compliments, but if I'm not wildly mistaken, it already *can* be used for its original purpose. E.g. `||example.org^$domain=example.com|~ Warning comment text`.

**Comment by Yuki2718** [comment]
> I see what you're thinking of, and I thank you for the compliments, but if I'm not wildly mistaken, it already _can_ be used for its original purpose. E.g. `||example.org^$domain=example.com|~ Warning comment text`.

Okay, I didn't know that. I still can't fully desert the idea of a dedicated option until declined, or unless uBO team guarantees there can be no detrimental effects in using domain option even with 16 bytes char such as 広告.

**Comment by DandelionSprout** [comment]
Now that you mention it, the `$domain` trick currently has zero support for non-ASCII letters, or for commas. So that's one thing that could theorethically be supported and improved on in `$message`.

**Comment by gorhill** [comment]
Trying to support messages is quite a task, let alone localized messages.

**Comment by curbengh** [comment]
> @Yuki2718
> Most users of my lists would have no idea about what to do if such a link is blocked
My understanding of this suggestion is that it can help user (and maintainer) of a merged blocklist (e.g. Energized Protection) to identify the upstream source of a filter, so that it's easy for them to report issue.

For my end, it's trivial for me to create a new blocklist with `$message` suffix, so I don't mind.

**Comment by Yuki2718** [comment]
> Trying to support messages is quite a task, let alone localized messages.

Then how about a new special comment, in addition to currently supported "! Title:" etc., one special comment for strict blocking per one list. Not as flexible as the original proposal (I'll need to list up all the possible reason, but for a dedicated anti-malware list it won't be a big issue) but given localized titles are correctly displayed, I guess this will solve the localization matter.

Ex.
```
! Homepage: http://example.com/
! Title: Foo Anti-phishing List
! Expires: 5 days
! Block-message: The site was blocked for phishing activity.
```

Maybe it can be loaded with Title but not shown as a title, instead only seen in a strict-blocked page with/under the title currently shown.

@curbengh Thx for the valuable comment, that's what I couldn't think of. My new suggestion above may not support this use case in merged lists, but anyway @gorhill will know what's the best to balance between time to spend and the benefit by the enhancement. Nice to know you can utilize the new option if adopted.

**Comment by gorhill** [comment]
> identify the upstream source of a filter

The list(s) containing the filter is already reported in the document-blocked page.

**Comment by gorhill** [comment]
>     ! Title: Foo Anti-phishing List
>     ! Expires: 5 days
>     ! Block-message: The site was blocked for phishing activity.

I don't see the message adding useful information, the title of the list is already telling us that the page was blocked because of a filter in "Foo Anti-phishing List".

**Comment by Yuki2718** [comment]
Uh, yes, in this case. And admittedly explaining all the cases in that way in a general list will not be trivial and may possibly confuse user.

**Comment by curbengh** [comment]
> https://github.com/uBlockOrigin/uBlock-issues/issues/1195#issuecomment-669906215
> identify the upstream source of a filter

I meant in a merged list, say I combined blocklists X and Y

```
! Title: Combined Blocklist
||badsite1.com^$message="Blocklist X"
||badsite2.com^$message="Blocklist Y"
```

then uBO can show "Combined Blocklist" and "Blocklist X" when a filter is triggered.

Currently blocklist maintainer could add a comment:

```
! Title: Combined Blocklist

! Blocklist X
||badsite1.com^

! Blocklist Y
||badsite2.com^
```

This way, maintainer still can easily find the source of a filter, user could also do it via "view content"; my understanding (could be different from OP's) of this `$message` suggestion is that uBO can directly show the source instead.

Personally, I find `! Blocklist X` approach is sufficient and hardly increase the file size of a merged blocklist.

**Comment by krystian3w** [comment]
Reddit idea to block malware pages with diffrent pop-up and have another switch:

https://pl.reddit.com/r/uBlockOrigin/comments/k1yyf7/feature_idea_feedback/

**Comment by Yuki2718** [comment]
Somehow I can't find with cursory search but IIRC there was similar suggestion on Github.

**Comment by garry-ut99** [comment]
Duplicate of https://github.com/gorhill/uBlock/issues/352 and https://github.com/uBlockOrigin/uBlock-issues/issues/475 which both were declined (and 475 was already a duplicate of 352).
<br>

> Yuki2718 : I'm impressed with clever use of `$domain=` by DandelionSprout https://github.com/DandelionSprout/adfilt/pull/81#issuecomment-655701790

And earlier by gwarser: https://github.com/gorhill/uBlock/issues/352#issuecomment-397467928

**Comment by krystian3w** [comment]
Until a spaces or missing dot+asterisk/regex deactivates the filter.

**Comment by gorhill** [comment]
The big issue in the proposal here is localization, and I don't see the feature request here as a duplicate of #475, as the other proposal is to have the message built into uBO, such that localization is not an issue.

**Comment by garry-ut99** [comment]
Everyone can have it's own definition of what duplicate is, your definition is from pure technical point of view, while mine is from a generic view, I see it like this: a very slightly different solution to an identical problem still falls into the duplicate category, as (a) it's about to solve the same problem, (b)  by the same solution (visually) - showing a message on a strict-blocked page, (c) which (the message) is being read from the same place - from a filter list, the only difference is, that instead of the message's text being read from a comment `!` it will be read from a filter itself, which is just several rows nearby the comment, the idea could be posted as a comment in the previous thread, but ok, next time I will use "duplicate or related" label to make everyone happy, as everyone can have it's own definition of what duplicate is, let's end this debate.

**Comment by gorhill** [comment]
I prefer to decline this.

**Comment by gorhill** [comment]
In light on [this discussion](https://github.com/uBlockOrigin/uBlock-discussions/discussions/771#discussioncomment-5551268), I reopen.

**Comment by gorhill** [comment]
`message=...` option added to latest dev build. Should be used everywhere `to=~...`.

This is a first step. Next, to decide how to render the message:
- as is, i.e. not different than how `to=~...` is currently rendered
- extract the message part from the filter text and render it separately

Also, possible future improvement: use message as key to lookup localized string:
- if a localized string is found, use it instead
- if not found, use message as is

So if a message pattern become common, we could decide to localize it, and no code change would be necessary to have it take effect.

**Comment by Yuki2718** [comment]
How to use `$message` currently? I tested with `||example.com^$doc,message=a` but it doesn't show the message except for the matched filter part as has been the case.

**Comment by DandelionSprout** [comment]
After the original declines and the closing of the "non-TLD $domain value" loopholes, I gradually landed on placing comments in e.g. `domain=~john_madden_aeiou.*` through wildcard and underscore loopholes that I discovered more than a year later independently of each other, and which seem to work pretty well.

In the event that `$message` sensationally (or at least it would be sensational in my eyes) does get implemented after all. I'm thinking we could just as well go all-in on it: `$message="John Madden aeiou"`. İ.e. allow fully regularly written text with spaces and everything. The text would be in quotation marks as what is customary in coding, but the quotation marks would not be shown in the displayed message.

**Comment by gorhill** [comment]
> but it doesn't show the message except for the matched filter part as has been the case.

Yes, it's what I alluded to when I wrote "Next, to decide how to render the message". Currently it's not worse than it what before message-wise, but filter-wise it's much better as we do not have to pollute the filters with pointless options, which is not ideal when converting to DNR.

About where I should render the message content somewhere else on the strictblock page, I am open to suggestions.

**Comment by Yuki2718** [comment]
>  On the downside, that AdGuard does not provide the name of the Filter (lists in uBO/ABP jargon).

Yes, I think it's bad and opened an issue: https://github.com/AdguardTeam/CoreLibs/issues/1847

**Comment by Yuki2718** [comment]
> How about this?

Just a personal opiniton, but I prefer the message displayed either on top of the page or below the blocked url.

**Comment by gorhill** [comment]
I am having second thought about `message=`. Implementing the above, I feel like `reason=` would be a more accurate name for the option. Thoughts?

**Comment by gorhill** [comment]
Ok I will set and update a list according to feedback (suggestions for user-facing text also welcome):

- `malicious` => "Reason: Malicious"
- `tracker` => "Reason: Tracker"

**Comment by DandelionSprout** [comment]
When it comes to messages and to get uBO users to read them, one thing that came to my mind to suggest is that such messages must be standard font and not monofont.

This is what it would look like with the current font (Using the F12 filetree to change the text, but lacked time to change the browser language to English in a hurry):
<details>

![Image](https://github.com/user-attachments/assets/1afe6a5e-6dbf-41ff-bd52-12121a4165c2)

</details>

This is what it would look like with the font-family CSS set to `"Segoe UI", "Calibri", "Verdana", "Ubuntu", "Bahnschrift", "Noto Sans CJK", "Malgun Gothic", sans-serif`, which in my case results in Segoe UI:

<details>

![Image](https://github.com/user-attachments/assets/7293c588-0b94-4e6f-9d81-4966f98b6d54)

</details>

There's probably many more fonts to add to such a hardcoded font list, but the more the better, in order to avoid users' pages from falling back to honestly terrifying fonts like MS PGothic for Japanese text, that one very pointy Cyrillic font I see from time to time on forums, or even odd-looking renders of Basic-2 Latin letters that I see on many newssites.

**Comment by gorhill** [comment]
With latest commit, `message` option has become `reason`, `message` won't work anymore. There are three built-in reasons which are localized:

- `malicious` => Malicious
- `tracker` => Tracker
- `disreputable` => Disreputable

Keep in mind that when using `reason` option, this potentially disables optimization of how filters are loaded in memory, i.e. a filter such as `||example.com^` is optimized by being stored in a trie, whereas `||example.com^$reason=tracker` will prevent this.

**Comment by Yuki2718** [comment]
> Keep in mind that when using `reason` option, this potentially disables optimization of how filters are loaded in memory, i.e. a filter such as `||example.com^` is optimized by being stored in a trie, whereas `||example.com^$reason=tracker` will prevent this.

No such downside for regex filters?

**Comment by gorhill** [comment]
These would be in a trie, there are multiple distinct ones for each primary options such as resource types (`script`, `xhr`, etc), domain types (no-party, `1p`, `3p`), block/allow, etc. You can see these tries in "SFNE: Dump" results, it's `FilterHostnameDict`.

**Comment by u-RraaLL** [comment]
Was there an issue publishing b8? The commit refers to it, but it's not in `/releases`, only in `/tags`..

**Comment by u-RraaLL** [comment]
BTW how will badfiltering work for filters with the `reason` option? ([Ref](https://github.com/uBlockOrigin/uBlock-discussions/discussions/833#discussioncomment-13543953))

Will `||discord.com^$badfilter` disable user's own `||discord.com^$reason="Self-control"` or should `badfilter` have an exact match, not to undermine possible user's choice?  `||discord.com^$reason="Self-control",badfilter`

**Comment by krystian3w** [comment]
It seems to work differently (if I'm not doing something wrong):

```adb
! win second
example.com$doc,reason="test-1"
example.com$doc,reason="test-2"
```


```adb
! win thrid
example.com$doc,reason="test-1"
example.com$doc,reason="test-2"
example.com$doc,frame
```

```adb
! win first
example.com$doc,reason="test-1"
example.com$doc,reason="test-2"
example.com$doc,frame,important
```

The further I check, the more I see the point of using a debugger or previewing what went into the comipled version of all filters. So for popular blocks lists there should be a common opinion when it is not a bug (so that different update intervals do not overwrite the behavior).

for `$doc,to=,reason=` (bypassing the URL test matching), looks the order is decisive.

**Comment by gorhill** [comment]
Whichever filter is hit first is undefined and whichever filter is hit first is the one which will be used for reporting.

**Comment by gorhill** [comment]
Ok I did navigate to `https://backpackmoments.com/fergana-valley-uzbekistan-full-guide/` and clicked the link "Rohat Hotel" in the text, and the filter `||jdoqocy.com/click-*&url=$urlskip=?url -uricomponent,reason=tracker` did not cause the reason for the blocking to be reported in the strict-block page.

I will make modification such that the reason, if not present in the block filter, will be also looked uo from the `urlskip` filter if present.

**Comment by Yuki2718** [comment]
~~@gorhill I tested `||adjust.com^$reason=tracker` but the display doesn't change when accessed `https://adjust.com/` with 1.67 on Chrome. What did I make wrong?~~
Okay, it seems I have to add `||adjust.com^$doc,badfilter`. Currently it only shows `Reason: Tracker`, but a bit more detailed explanation like the one in Brave is desirable.

<img width="1541" height="902" alt="Image" src="https://github.com/user-attachments/assets/0bfc84c4-7e52-48f2-894a-75e37eb7b804" />
